File: java-server/src/main/java/com/chatsphere/server/controller/MessageController.java
package com.chatsphere.server.controller;

import com.chatsphere.server.document.Message;
import com.chatsphere.server.entity.DirectConversation;
import com.chatsphere.server.entity.MessageRequest;
import com.chatsphere.server.entity.User;
import com.chatsphere.server.repository.DirectConversationRepository;
import com.chatsphere.server.repository.MessageRequestRepository;
import com.chatsphere.server.repository.MessageRepository;
import com.chatsphere.server.repository.UserRepository;
import com.chatsphere.server.security.JwtAuthenticationFilter.UserPrincipal;
import com.corundumstudio.socketio.SocketIOServer;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;
import java.io.File;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.time.Instant;
import java.time.LocalDateTime;
import java.time.ZoneId;
import java.util.*;
import java.util.stream.Collectors;

@RestController
@RequestMapping("/api/messages")
public class MessageController {

    private final MessageRepository messageRepository;
    private final DirectConversationRepository directConversationRepository;
    private final MessageRequestRepository messageRequestRepository;
    private final UserRepository userRepository;
    private final SocketIOServer socketServer;

    public MessageController(MessageRepository messageRepository,
                             DirectConversationRepository directConversationRepository,
                             MessageRequestRepository messageRequestRepository,
                             UserRepository userRepository,
                             SocketIOServer socketServer) {
        this.messageRepository = messageRepository;
        this.directConversationRepository = directConversationRepository;
        this.messageRequestRepository = messageRequestRepository;
        this.userRepository = userRepository;
        this.socketServer = socketServer;
    }

    private UserPrincipal getAuthenticatedUser() {
        return (UserPrincipal) SecurityContextHolder.getContext().getAuthentication().getPrincipal();
    }

    // --- 1. GET MESSAGES (Room or DM) ---
    @GetMapping("/{roomId}")
    public ResponseEntity<?> getMessages(@PathVariable String roomId) {
        Integer userId = getAuthenticatedUser().getId();
        boolean isDirect = roomId.startsWith("dm_");
        List<Message> messages;

        if (isDirect) {
            String convId = roomId.replace("dm_", "");
            Optional<DirectConversation> convOpt = directConversationRepository.findById(Integer.parseInt(convId));
            
            Instant clearedAt = null;
            if (convOpt.isPresent()) {
                DirectConversation conv = convOpt.get();
                LocalDateTime dt = conv.getUser1Id().equals(userId) ? conv.getUser1ClearedAt() : conv.getUser2ClearedAt();
                if (dt != null) {
                    clearedAt = dt.atZone(ZoneId.systemDefault()).toInstant();
                }
            }

            if (clearedAt != null) {
                messages = messageRepository.findByConversationIdAndCreatedAtGreaterThanOrderByCreatedAtAsc(convId, clearedAt);
            } else {
                messages = messageRepository.findByConversationIdOrderByCreatedAtAsc(convId);
            }
        } else {
            messages = messageRepository.findByRoomIdOrderByCreatedAtAsc(roomId);
        }

        // Limit to last 100 if list is too large
        if (messages.size() > 100) {
            messages = messages.subList(messages.size() - 100, messages.size());
        }

        return ResponseEntity.ok(messages);
    }

    // --- 2. UPLOAD MEDIA ---
    @PostMapping("/upload")
    public ResponseEntity<?> uploadMedia(@RequestParam("file") MultipartFile file) {
        if (file.isEmpty()) {
            return ResponseEntity.badRequest().body(Map.of("message", "No file uploaded"));
        }

        try {
            String uploadsPath = new File("../server/uploads/").getAbsolutePath();
            File destDir = new File(uploadsPath);
            if (!destDir.exists()) {
                destDir.mkdirs();
            }

            String origName = file.getOriginalFilename();
            String extension = origName != null && origName.contains(".") ? origName.substring(origName.lastIndexOf(".")) : "";
            String newFileName = System.currentTimeMillis() + "_" + java.util.UUID.randomUUID().toString().substring(0, 8) + extension;
            
            Path path = Paths.get(uploadsPath, newFileName);
            Files.write(path, file.getBytes());

            String fileUrl = "/uploads/" + newFileName;

            return ResponseEntity.ok(Map.of(
                "fileUrl", fileUrl,
                "fileName", origName != null ? origName : newFileName,
                "fileType", file.getContentType() != null ? file.getContentType() : "",
                "size", file.getSize()
            ));
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(Map.of("message", "Error uploading media: " + e.getMessage()));
        }
    }

    // --- 3. GET MEDIA MESSAGES FOR PANEL ---
    @GetMapping("/{conversationId}/media")
    public ResponseEntity<?> getMediaMessages(@PathVariable String conversationId) {
        String cleanId = conversationId.replace("dm_", "");
        String roomId = "dm_" + cleanId;
        Pageable limit = PageRequest.of(0, 50, Sort.by("createdAt").descending());
        
        List<Message> mediaMsgs = messageRepository.findMediaMessages(cleanId, roomId, limit);
        
        List<Map<String, Object>> formatted = mediaMsgs.stream().map(m -> Map.of(
            "fileUrl", m.getFileUrl() == null ? "" : m.getFileUrl(),
            "fileName", m.getFileName() == null ? "" : m.getFileName(),
            "fileType", m.getFileType() == null ? "" : m.getFileType(),
            "createdAt", m.getCreatedAt(),
            "senderId", m.getSenderId()
        )).collect(Collectors.toList());

        return ResponseEntity.ok(formatted);
    }

    // --- 4. TOGGLE REACTION ---
    @PostMapping("/{messageId}/react")
    public ResponseEntity<?> toggleReaction(@PathVariable String messageId, @RequestBody Map<String, String> request) {
        String emoji = request.get("emoji");
        String username = request.get("username");
        Integer userId = getAuthenticatedUser().getId();

        if (emoji == null || emoji.trim().isEmpty()) {
            return ResponseEntity.badRequest().body(Map.of("message", "Emoji is required"));
        }

        Optional<Message> msgOpt = messageRepository.findById(messageId);
        if (msgOpt.isEmpty()) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND).body(Map.of("message", "Message not found"));
        }

        Message message = msgOpt.get();
        
        // Find reaction
        Message.Reaction targetReaction = null;
        for (Message.Reaction r : message.getReactions()) {
            if (r.getEmoji().equals(emoji)) {
                targetReaction = r;
                break;
            }
        }

        if (targetReaction != null) {
            // Check if user already reacted
            Message.ReactionUser targetUser = null;
            for (Message.ReactionUser ru : targetReaction.getUsers()) {
                if (ru.getUserId().toString().equals(userId.toString())) {
                    targetUser = ru;
                    break;
                }
            }

            if (targetUser != null) {
                // Remove reaction
                targetReaction.getUsers().remove(targetUser);
                if (targetReaction.getUsers().isEmpty()) {
                    message.getReactions().remove(targetReaction);
                }
            } else {
                // Add user to existing emoji
                targetReaction.getUsers().add(new Message.ReactionUser(userId, username));
            }
        } else {
            // Create new reaction
            Message.Reaction r = new Message.Reaction();
            r.setEmoji(emoji);
            r.getUsers().add(new Message.ReactionUser(userId, username));
            message.getReactions().add(r);
        }

        Message savedMsg = messageRepository.save(message);

        // Emit socket event to the room
        if (socketServer != null) {
            socketServer.getRoomOperations(message.getRoomId()).sendEvent("messageReacted", savedMsg);
        }

        return ResponseEntity.ok(savedMsg);
    }

    // --- 5. CLEAR MESSAGES ---
    @DeleteMapping("/{conversationId}")
    @Transactional
    public ResponseEntity<?> clearMessages(@PathVariable String conversationId) {
        Integer userId = getAuthenticatedUser().getId();
        String cleanId = conversationId.replace("dm_", "");

        Optional<DirectConversation> convOpt = directConversationRepository.findById(Integer.parseInt(cleanId));
        if (convOpt.isEmpty()) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND).body(Map.of("message", "Conversation not found"));
        }

        DirectConversation conv = convOpt.get();
        if (conv.getUser1Id().equals(userId)) {
            conv.setUser1ClearedAt(LocalDateTime.now());
        } else {
            conv.setUser2ClearedAt(LocalDateTime.now());
        }

        directConversationRepository.save(conv);

        // Emit clear message event to this user's socket room
        if (socketServer != null) {
            socketServer.getRoomOperations("user_" + userId).sendEvent("chatCleared", Map.of("roomId", conversationId));
        }

        return ResponseEntity.ok(Map.of("message", "Chat cleared successfully"));
    }

    // --- 6. CREATE DIRECT MESSAGE REQUEST ---
    @PostMapping("/request")
    public ResponseEntity<?> createRequest(@RequestBody Map<String, Object> request) {
        Object toUserIdObj = request.get("toUserId");
        if (toUserIdObj == null) {
            return ResponseEntity.badRequest().body(Map.of("message", "toUserId is required"));
        }

        Integer fromUserId = getAuthenticatedUser().getId();
        String fromUsername = getAuthenticatedUser().getUsername();
        Integer toUserId = Integer.valueOf(toUserIdObj.toString());

        if (fromUserId.equals(toUserId)) {
            return ResponseEntity.badRequest().body(Map.of("message", "Cannot message yourself"));
        }

        // Check if conversation already exists
        Optional<DirectConversation> existing = directConversationRepository.findConversationBetween(fromUserId, toUserId);
        if (existing.isPresent()) {
            return ResponseEntity.ok(Map.of("message", "Conversation exists", "conversationId", existing.get().getId()));
        }

        // Check if request already pending
        Optional<MessageRequest> pending = messageRequestRepository.findByFromUserIdAndToUserIdAndStatus(fromUserId, toUserId, "pending");
        if (pending.isPresent()) {
            return ResponseEntity.badRequest().body(Map.of("message", "Request already sent"));
        }

        MessageRequest req = new MessageRequest();
        req.setFromUserId(fromUserId);
        req.setFromUsername(fromUsername);
        req.setToUserId(toUserId);
        MessageRequest savedReq = messageRequestRepository.save(req);

        if (socketServer != null) {
            socketServer.getRoomOperations("user_" + toUserId).sendEvent("incomingRequest", savedReq);
        }

        return ResponseEntity.status(HttpStatus.CREATED).body(Map.of("message", "Message request sent successfully"));
    }

    // --- 7. GET PENDING REQUESTS ---
    @GetMapping("/requests/pending")
    public ResponseEntity<?> getPendingRequests() {
        Integer userId = getAuthenticatedUser().getId();
        List<MessageRequest> requests = messageRequestRepository.findByToUserIdAndStatus(userId, "pending");
        return ResponseEntity.ok(requests);
    }

    // --- 8. ACCEPT REQUEST ---
    @PutMapping("/request/{id}/accept")
    @Transactional
    public ResponseEntity<?> acceptRequest(@PathVariable Integer id) {
        Integer userId = getAuthenticatedUser().getId();

        Optional<MessageRequest> reqOpt = messageRequestRepository.findById(id);
        if (reqOpt.isEmpty()) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND).body(Map.of("message", "Request not found"));
        }

        MessageRequest req = reqOpt.get();
        if (!req.getToUserId().equals(userId)) {
            return ResponseEntity.status(HttpStatus.FORBIDDEN).body(Map.of("message", "Unauthorized"));
        }

        req.setStatus("accepted");
        messageRequestRepository.save(req);

        Integer user1 = req.getFromUserId();
        Integer user2 = userId;

        Optional<DirectConversation> existing = directConversationRepository.findConversationBetween(user1, user2);
        Integer conversationId;
        if (existing.isPresent()) {
            conversationId = existing.get().getId();
        } else {
            DirectConversation dc = new DirectConversation();
            dc.setUser1Id(user1);
            dc.setUser2Id(user2);
            DirectConversation savedDc = directConversationRepository.save(dc);
            conversationId = savedDc.getId();
        }

        if (socketServer != null) {
            Map<String, Object> payload = Map.of(
                "conversationId", conversationId,
                "requestId", id
            );
            socketServer.getRoomOperations("user_" + user1).sendEvent("requestAccepted", payload);
            socketServer.getRoomOperations("user_" + user2).sendEvent("requestAccepted", payload);
        }

        return ResponseEntity.ok(Map.of(
            "message", "Request accepted",
            "conversation", Map.of(
                "id", conversationId,
                "user1_id", user1,
                "user2_id", user2,
                "otherUsername", req.getFromUsername(),
                "otherUserId", req.getFromUserId()
            )
        ));
    }

    // --- 9. BLOCK REQUEST ---
    @PutMapping("/request/{id}/block")
    public ResponseEntity<?> blockRequest(@PathVariable Integer id) {
        Optional<MessageRequest> reqOpt = messageRequestRepository.findById(id);
        if (reqOpt.isEmpty()) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND).body(Map.of("message", "Request not found"));
        }

        MessageRequest req = reqOpt.get();
        req.setStatus("blocked");
        messageRequestRepository.save(req);

        if (socketServer != null) {
            socketServer.getRoomOperations("user_" + req.getToUserId()).sendEvent("requestBlocked", Map.of("requestId", id));
            socketServer.getRoomOperations("user_" + req.getFromUserId()).sendEvent("requestBlocked", Map.of("requestId", id));
        }

        return ResponseEntity.ok(Map.of("message", "User blocked"));
    }

    // --- 10. GET DIRECT CONVERSATIONS ---
    @GetMapping("/direct/conversations")
    public ResponseEntity<?> getDirectConversations() {
        Integer userId = getAuthenticatedUser().getId();

        List<DirectConversation> convs = directConversationRepository.findByUser1IdOrUser2Id(userId, userId);

        List<Map<String, Object>> formatted = convs.stream().map(c -> {
            Integer otherUserId = c.getUser1Id().equals(userId) ? c.getUser2Id() : c.getUser1Id();
            User otherUser = userRepository.findById(otherUserId).orElse(null);
            
            if (otherUser == null) return null;

            Map<String, Object> map = new HashMap<>();
            map.put("conversationId", c.getId());
            map.put("otherUserId", otherUser.getId());
            map.put("name", otherUser.getUsername());
            map.put("profile_pic", otherUser.getProfilePic() == null ? "" : otherUser.getProfilePic());
            map.put("user1_id", c.getUser1Id());
            map.put("user2_id", c.getUser2Id());

            LocalDateTime dt = c.getUser1Id().equals(userId) ? c.getUser1ClearedAt() : c.getUser2ClearedAt();
            Instant clearedAt = dt == null ? null : dt.atZone(ZoneId.systemDefault()).toInstant();

            // Find last message
            Optional<Message> lastMsgOpt;
            String cleanId = String.valueOf(c.getId());
            String roomId = "dm_" + cleanId;
            Pageable oneSortedDesc = PageRequest.of(0, 1, Sort.by(Sort.Direction.DESC, "createdAt"));

            if (clearedAt != null) {
                List<Message> msgs = messageRepository.findFirstByConversationIdOrRoomIdAndCreatedAtGreaterThanOrderByCreatedAtDesc(cleanId, roomId, clearedAt, oneSortedDesc);
                lastMsgOpt = msgs.isEmpty() ? Optional.empty() : Optional.of(msgs.get(0));
            } else {
                List<Message> msgs = messageRepository.findFirstByConversationIdOrRoomIdOrderByCreatedAtDesc(cleanId, roomId, oneSortedDesc);
                lastMsgOpt = msgs.isEmpty() ? Optional.empty() : Optional.of(msgs.get(0));
            }

            if (lastMsgOpt.isPresent()) {
                Message lastMsg = lastMsgOpt.get();
                map.put("lastMessage", Map.of(
                    "content", lastMsg.getContent(),
                    "senderId", lastMsg.getSenderId(),
                    "status", lastMsg.getStatus() == null ? "sent" : lastMsg.getStatus(),
                    "createdAt", lastMsg.getCreatedAt(),
                    "type", lastMsg.getType(),
                    "fileName", lastMsg.getFileName() == null ? "" : lastMsg.getFileName(),
                    "isEncrypted", lastMsg.getIsEncrypted(),
                    "encryptedKey", lastMsg.getEncryptedKey() == null ? "" : lastMsg.getEncryptedKey(),
                    "senderEncryptedKey", lastMsg.getSenderEncryptedKey() == null ? "" : lastMsg.getSenderEncryptedKey(),
                    "iv", lastMsg.getIv() == null ? "" : lastMsg.getIv()
                ));
            } else {
                map.put("lastMessage", null);
            }

            return map;
        }).filter(Objects::nonNull).collect(Collectors.toList());

        // Sort by last message date
        formatted.sort((a, b) -> {
            Map<?, ?> msgA = (Map<?, ?>) a.get("lastMessage");
            Map<?, ?> msgB = (Map<?, ?>) b.get("lastMessage");
            Instant tA = msgA == null ? Instant.ofEpochMilli(0) : (Instant) msgA.get("createdAt");
            Instant tB = msgB == null ? Instant.ofEpochMilli(0) : (Instant) msgB.get("createdAt");
            return tB.compareTo(tA);
        });

        return ResponseEntity.ok(formatted);
    }

    // --- 11. CHECK DIRECT CONVERSATION ---
    @GetMapping("/direct/{userId}")
    public ResponseEntity<?> checkDirectConversation(@PathVariable Integer userId) {
        Integer currentUserId = getAuthenticatedUser().getId();
        Optional<DirectConversation> existing = directConversationRepository.findConversationBetween(currentUserId, userId);
        if (existing.isPresent()) {
            return ResponseEntity.ok(Map.of("exists", true, "conversationId", existing.get().getId()));
        }
        return ResponseEntity.ok(Map.of("exists", false));
    }
}

File: server/controllers/messageController.js
const Message = require('../models/Message');
const db = require('../config/db');
const fs = require('fs');

// ── GET /api/messages/:roomId ─────────────────────────────────────────────────
exports.getMessages = async (req, res) => {
  try {
    const { roomId } = req.params;
    const userId = req.user.id;
    const isDirect = roomId.toString().startsWith('dm_');
    let messages;

    if (isDirect) {
      const convId = roomId.toString().replace('dm_', '');
      
      // Ensure columns exist (Migration)
      try {
        const [columns] = await db.query('SHOW COLUMNS FROM DirectConversations');
        const columnNames = columns.map(c => c.Field);
        if (!columnNames.includes('user1_cleared_at')) {
          await db.query('ALTER TABLE DirectConversations ADD COLUMN user1_cleared_at TIMESTAMP NULL DEFAULT NULL');
        }
        if (!columnNames.includes('user2_cleared_at')) {
          await db.query('ALTER TABLE DirectConversations ADD COLUMN user2_cleared_at TIMESTAMP NULL DEFAULT NULL');
        }
      } catch (e) {
        console.error('Migration error in getMessages:', e.message);
      }

      // Fetch cleared_at for the current user
      const [convs] = await db.query(
        'SELECT user1_id, user2_id, user1_cleared_at, user2_cleared_at FROM DirectConversations WHERE id = ?',
        [convId]
      );

      let clearedAt = null;
      if (convs.length > 0) {
        const conv = convs[0];
        if (conv.user1_id === userId) clearedAt = conv.user1_cleared_at;
        else if (conv.user2_id === userId) clearedAt = conv.user2_cleared_at;
      }

      const query = { conversationId: convId };
      if (clearedAt) {
        query.createdAt = { $gt: new Date(clearedAt) };
      }

      messages = await Message.find(query).sort({ createdAt: 1 }).limit(100);
    } else {
      messages = await Message.find({ roomId }).sort({ createdAt: 1 }).limit(100);
    }
    res.status(200).json(messages);
  } catch (error) {
    console.error('Error fetching messages:', error);
    res.status(500).json({ message: 'Server error while fetching messages' });
  }
};

// ── POST /api/messages/upload ─────────────────────────────────────────────────
exports.uploadMedia = async (req, res) => {
  try {
    if (!req.file) return res.status(400).json({ message: 'No file uploaded' });
    const fileUrl = `/uploads/${req.file.filename}`;
    res.status(200).json({
      fileUrl,
      fileName: req.file.originalname,
      fileType: req.file.mimetype,
      size: req.file.size
    });
  } catch (error) {
    console.error('Error uploading media:', error);
    res.status(500).json({ message: 'Server error while uploading media' });
  }
};

// ── GET /api/messages/:conversationId/media ───────────────────────────────────
// Returns only media-type messages for the profile panel media grid
exports.getMediaMessages = async (req, res) => {
  try {
    const { conversationId } = req.params;
    const cleanId = conversationId.replace('dm_', '');

    const mediaMessages = await Message.find({
      $or: [{ conversationId: cleanId }, { roomId: `dm_${cleanId}` }],
      type: 'media'
    })
      .sort({ createdAt: -1 })
      .limit(50)
      .select('fileUrl fileName fileType createdAt senderId')
      .lean();

    res.status(200).json(mediaMessages);
  } catch (error) {
    console.error('Error fetching media messages:', error);
    res.status(500).json({ message: 'Server error while deleting chat' });
  }
};

// ── POST /api/messages/:messageId/react ───────────────────────────────────────
exports.toggleReaction = async (req, res) => {
  try {
    const { messageId } = req.params;
    const { emoji, username } = req.body;
    const userId = req.user.id;

    if (!emoji) return res.status(400).json({ message: 'Emoji is required' });

    const message = await Message.findById(messageId);
    if (!message) return res.status(404).json({ message: 'Message not found' });

    // Find if reaction with this emoji exists
    let reactionIndex = message.reactions.findIndex(r => r.emoji === emoji);

    if (reactionIndex > -1) {
      // Check if user already reacted with this emoji
      const userIndex = message.reactions[reactionIndex].users.findIndex(u => u.userId === userId);
      
      if (userIndex > -1) {
        // Remove reaction
        message.reactions[reactionIndex].users.splice(userIndex, 1);
        // If no users left for this emoji, remove the emoji entirely
        if (message.reactions[reactionIndex].users.length === 0) {
          message.reactions.splice(reactionIndex, 1);
        }
      } else {
        // Add user to existing emoji
        message.reactions[reactionIndex].users.push({ userId, username });
      }
    } else {
      // Create new emoji reaction
      message.reactions.push({
        emoji,
        users: [{ userId, username }]
      });
    }

    await message.save();

    await message.save();

    const io = req.app.get('io');
    if (io) {
      io.to(message.roomId).emit('messageReacted', message);
    }

    res.status(200).json(message);
  } catch (error) {
    console.error('Error toggling reaction:', error);
    res.status(500).json({ message: 'Server error while toggling reaction' });
  }
};

// ── DELETE /api/messages/:conversationId ──────────────────────────────────────
// "Clears" chat for the requesting user only by updating cleared_at timestamp in MySQL
exports.clearMessages = async (req, res) => {
  try {
    const { conversationId } = req.params;
    const userId = req.user.id;
    const cleanId = conversationId.replace('dm_', '');

    // Ensure columns exist (Migration)
    try {
      await db.query('ALTER TABLE DirectConversations ADD COLUMN user1_cleared_at TIMESTAMP NULL DEFAULT NULL');
      await db.query('ALTER TABLE DirectConversations ADD COLUMN user2_cleared_at TIMESTAMP NULL DEFAULT NULL');
    } catch (_) { /* ignore if already exist */ }

    // Find which user is requesting the clear
    const [convs] = await db.query(
      'SELECT user1_id, user2_id FROM DirectConversations WHERE id = ?',
      [cleanId]
    );

    if (convs.length === 0) return res.status(404).json({ message: 'Conversation not found' });

    const conv = convs[0];
    const column = (conv.user1_id === userId) ? 'user1_cleared_at' : 'user2_cleared_at';

    await db.query(
      `UPDATE DirectConversations SET ${column} = NOW() WHERE id = ?`,
      [cleanId]
    );

    res.status(200).json({ message: 'Chat cleared for you successfully' });
  } catch (error) {
    console.error('Error clearing messages:', error);
    res.status(500).json({ message: 'Server error while clearing messages' });
  }
};

// ── POST /api/messages/request ────────────────────────────────────────────────
exports.createRequest = async (req, res) => {
  try {
    const { toUserId } = req.body;
    const fromUserId = req.user.id;
    const fromUsername = req.user.username;

    if (Number(toUserId) === fromUserId) {
      return res.status(400).json({ message: 'Cannot message yourself' });
    }

    // Check if conversation already exists
    const [existingConv] = await db.query(
      `SELECT * FROM DirectConversations 
       WHERE (user1_id = ? AND user2_id = ?) OR (user1_id = ? AND user2_id = ?)`,
      [fromUserId, toUserId, toUserId, fromUserId]
    );
    if (existingConv.length > 0) {
      return res.status(200).json({ message: 'Conversation exists', conversationId: existingConv[0].id });
    }

    // Check if request already sent
    const [existingReq] = await db.query(
      `SELECT * FROM MessageRequests WHERE from_user_id = ? AND to_user_id = ? AND status = 'pending'`,
      [fromUserId, toUserId]
    );
    if (existingReq.length > 0) {
      return res.status(400).json({ message: 'Request already sent' });
    }

    await db.query(
      `INSERT INTO MessageRequests (from_user_id, from_username, to_user_id) VALUES (?, ?, ?)`,
      [fromUserId, fromUsername, toUserId]
    );

    res.status(201).json({ message: 'Message request sent successfully' });
  } catch (error) {
    console.error('Error creating request:', error);
    res.status(500).json({ message: 'Server error' });
  }
};

// ── GET /api/messages/requests/pending ───────────────────────────────────────
exports.getRequests = async (req, res) => {
  try {
    const userId = req.user.id;
    const [requests] = await db.query(
      `SELECT * FROM MessageRequests WHERE to_user_id = ? AND status = 'pending'`,
      [userId]
    );
    res.status(200).json(requests);
  } catch (error) {
    console.error('Error fetching requests:', error);
    res.status(500).json({ message: 'Server error' });
  }
};

// ── PUT /api/messages/request/:id/accept ─────────────────────────────────────
exports.acceptRequest = async (req, res) => {
  try {
    const { id } = req.params;
    const userId = req.user.id;

    const [requests] = await db.query(`SELECT * FROM MessageRequests WHERE id = ?`, [id]);
    if (requests.length === 0) return res.status(404).json({ message: 'Request not found' });

    const request = requests[0];
    if (request.to_user_id !== userId) return res.status(403).json({ message: 'Unauthorized' });

    await db.query(`UPDATE MessageRequests SET status = 'accepted' WHERE id = ?`, [id]);

    const user1 = request.from_user_id;
    const user2 = userId;

    const [existingConv] = await db.query(
      `SELECT * FROM DirectConversations 
       WHERE (user1_id = ? AND user2_id = ?) OR (user1_id = ? AND user2_id = ?)`,
      [user1, user2, user2, user1]
    );

    let conversationId;
    if (existingConv.length > 0) {
      conversationId = existingConv[0].id;
    } else {
      const [result] = await db.query(
        `INSERT INTO DirectConversations (user1_id, user2_id) VALUES (?, ?)`,
        [user1, user2]
      );
      conversationId = result.insertId;
    }

    res.status(200).json({
      message: 'Request accepted',
      conversation: {
        id: conversationId,
        user1_id: user1,
        user2_id: user2,
        otherUsername: request.from_username,
        otherUserId: request.from_user_id
      }
    });
  } catch (error) {
    console.error('Error accepting request:', error);
    res.status(500).json({ message: 'Server error' });
  }
};

// ── PUT /api/messages/request/:id/block ──────────────────────────────────────
exports.blockRequest = async (req, res) => {
  try {
    const { id } = req.params;
    await db.query(`UPDATE MessageRequests SET status = 'blocked' WHERE id = ?`, [id]);
    res.status(200).json({ message: 'User blocked' });
  } catch (error) {
    console.error('Error blocking request:', error);
    res.status(500).json({ message: 'Server error' });
  }
};

// ── GET /api/messages/direct/conversations ────────────────────────────────────
exports.getDirectConversations = async (req, res) => {
  try {
    const userId = req.user.id;

    // Ensure columns exist (Migration)
    try {
      const [columns] = await db.query('SHOW COLUMNS FROM DirectConversations');
      const columnNames = columns.map(c => c.Field);
      if (!columnNames.includes('user1_cleared_at')) {
        await db.query('ALTER TABLE DirectConversations ADD COLUMN user1_cleared_at TIMESTAMP NULL DEFAULT NULL');
      }
      if (!columnNames.includes('user2_cleared_at')) {
        await db.query('ALTER TABLE DirectConversations ADD COLUMN user2_cleared_at TIMESTAMP NULL DEFAULT NULL');
      }
    } catch (e) {
      console.error('Migration error:', e.message);
    }

    // Try with BlockedUsers JOIN — falls back if table doesn't exist yet
    let convos;
    try {
      [convos] = await db.query(`
        SELECT dc.id as conversationId, u.id as otherUserId, u.username as name, u.profile_pic,
               dc.user1_id, dc.user2_id, dc.user1_cleared_at, dc.user2_cleared_at
        FROM DirectConversations dc
        JOIN Users u ON (u.id = dc.user1_id OR u.id = dc.user2_id) AND u.id != ?
        LEFT JOIN BlockedUsers bu ON bu.blocker_id = ? AND bu.blocked_id = u.id
        WHERE (dc.user1_id = ? OR dc.user2_id = ?) AND bu.id IS NULL
      `, [userId, userId, userId, userId]);
    } catch (_) {
      // BlockedUsers table not created yet — query without it
      [convos] = await db.query(`
        SELECT dc.id as conversationId, u.id as otherUserId, u.username as name, u.profile_pic,
               dc.user1_id, dc.user2_id, dc.user1_cleared_at, dc.user2_cleared_at
        FROM DirectConversations dc
        JOIN Users u ON (u.id = dc.user1_id OR u.id = dc.user2_id) AND u.id != ?
        WHERE dc.user1_id = ? OR dc.user2_id = ?
      `, [userId, userId, userId]);
    }

    const conversationsWithLastMessage = await Promise.all(convos.map(async (convo) => {
      const convIdStr = String(convo.conversationId);
      const roomIdStr = `dm_${convIdStr}`;

      // Handle user1 vs user2 cleared_at logic from the query result
      // Note: convo should now include user1_cleared_at and user2_cleared_at
      let clearedAt = null;
      if (convo.user1_id === userId) clearedAt = convo.user1_cleared_at;
      else if (convo.user2_id === userId) clearedAt = convo.user2_cleared_at;

      const query = {
        $or: [{ conversationId: convIdStr }, { roomId: roomIdStr }]
      };
      
      if (clearedAt) {
        query.createdAt = { $gt: new Date(clearedAt) };
      }

      const lastMessage = await Message.findOne(query).sort({ createdAt: -1 }).select('content senderId status createdAt type fileName isEncrypted encryptedKey senderEncryptedKey iv').lean();
      return { ...convo, lastMessage: lastMessage || null };
    }));

    conversationsWithLastMessage.sort((a, b) => {
      const timeA = a.lastMessage ? new Date(a.lastMessage.createdAt) : new Date(0);
      const timeB = b.lastMessage ? new Date(b.lastMessage.createdAt) : new Date(0);
      return timeB - timeA;
    });

    res.status(200).json(conversationsWithLastMessage);
  } catch (error) {
    console.error('Error fetching conversations:', error);
    res.status(500).json({ message: 'Server error' });
  }
};


// ── GET /api/messages/direct/:userId ─────────────────────────────────────────
exports.checkDirectConversation = async (req, res) => {
  try {
    const currentUserId = req.user.id;
    const { userId: otherUserId } = req.params;

    const [existingConv] = await db.query(
      `SELECT * FROM DirectConversations 
       WHERE (user1_id = ? AND user2_id = ?) OR (user1_id = ? AND user2_id = ?)`,
      [currentUserId, otherUserId, otherUserId, currentUserId]
    );

    if (existingConv.length > 0) {
      return res.status(200).json({ exists: true, conversationId: existingConv[0].id });
    }
    res.status(200).json({ exists: false });
  } catch (error) {
    console.error('Error checking conversation:', error);
    res.status(500).json({ message: 'Server error' });
  }
};
